From 84234d7fea654f7318c51df51aa4f1487072c0e1 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Thu, 6 Jun 2019 15:55:04 +0200
Subject: [PATCH] TRITON-1325 node-triton fwrule support for cloud firewall
 logging Reviewed by: Mike Gerdts <mike.gerdts@joyent.com> Reviewed by: Trent
 Mick <trentm@gmail.com> Approved by: Trent Mick <trentm@gmail.com>

---
 CHANGES.md                           |  40 ++++++++
 lib/cloudapi2.js                     |   5 +
 lib/do_fwrule/do_create.js           |  19 +++-
 lib/do_fwrule/do_list.js             |   6 +-
 lib/do_fwrule/do_update.js           |   7 +-
 lib/do_instance/do_fwrule/do_list.js |   6 +-
 lib/do_instance/do_fwrule/index.js   |   2 +
 lib/do_instance/do_fwrules.js        |   2 +
 lib/tritonapi.js                     |  10 +-
 package.json                         |   2 +-
 test/integration/cli-fwrules.test.js | 145 +++++++++++++++++++--------
 11 files changed, 184 insertions(+), 60 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index de8ed0c..0c57a45 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -6,6 +6,46 @@ Known issues:
 
 ## not yet released
 
+## 7.2.0
+
+- [TRITON-1325] node-triton fwrule support for cloud firewall logging.
+  Firewall rules will now include the attribute `log (Boolean)`. When
+  true, the system will log new TCP connections or new other-protocol
+  sessions matching the rules.
+  [RFD 163](https://github.com/joyent/rfd/tree/master/rfd/0163)
+
+  This change modifies the default output of the `triton inst fwrules`
+  and `triton fwrule list` adding a `LOG` column to the default output.
+
+  For example, for a set of rules that previously were as follows:
+
+        SHORTID   ENABLED  GLOBAL  RULE
+        285d7f76  false    -       FROM any TO vm efe45825-4c0d-48f5-d62c-c5a50433fad1 BLOCK tcp PORT 666
+        4ef987de  true     -       FROM subnet 10.99.99.0/24 TO vm 3a2b9998-965d-c4ab-d952-eb2802f8d6b9 ALLOW tcp PORT all
+        44eae6bb  true     -       FROM subnet 10.99.99.0/24 TO vm efe45825-4c0d-48f5-d62c-c5a50433fad1 ALLOW tcp PORT all
+
+  The new output will be:
+
+        SHORTID   ENABLED  GLOBAL  LOG   RULE 
+        285d7f76  false    -       true  FROM any TO vm efe45825-4c0d-48f5-d62c-c5a50433fad1 BLOCK tcp PORT 666
+        4ef987de  true     -       true  FROM subnet 10.99.99.0/24 TO vm 3a2b9998-965d-c4ab-d952-eb2802f8d6b9 ALLOW tcp PORT all
+        44eae6bb  true     -       true  FROM subnet 10.99.99.0/24 TO vm efe45825-4c0d-48f5-d62c-c5a50433fad1 ALLOW tcp PORT all
+
+  The `log (Boolean)` field will be returned when a single firewall rule is
+  retrieved using `fwrule get`:
+
+        fwrule get 44eae6bb
+        {
+            "id": "44eae6bb-337f-45ba-8ff9-dddcd46e5918",
+            "rule": "FROM subnet 10.99.99.0/24 TO vm efe45825-4c0d-48f5-d62c-c5a50433fad1 ALLOW tcp PORT all",
+            "enabled": true,
+            "log": true
+        }
+
+  The sub-command `fwrule create` will include the new `-l|--log` option for
+  rule creation and the `log` value has been added to the list of fields which
+  can be updated using `fwrule update`.
+
 ## 7.1.1
 
 - [joyent/node-triton#169] Fix `triton rbac ...` commands that were
diff --git a/lib/cloudapi2.js b/lib/cloudapi2.js
index 965c124..932d94c 100644
--- a/lib/cloudapi2.js
+++ b/lib/cloudapi2.js
@@ -2081,6 +2081,7 @@ function waitForNicStates(opts, cb) {
  * @param {Object} options object containing:
  *      - {String} rule (required) the fwrule text.
  *      - {Boolean} enabled (optional) default to false.
+ *      - {Boolean} log (optional) default to false.
  *      - {String} description (optional)
  * @param {Function} callback of the form f(err, fwrule, res).
  */
@@ -2090,6 +2091,7 @@ function createFirewallRule(opts, cb) {
     assert.string(opts.rule, 'opts.rule');
     assert.optionalString(opts.description, 'opts.description');
     assert.optionalBool(opts.enabled, 'opts.enabled');
+    assert.optionalBool(opts.log, 'opts.log');
 
     var data = {};
     Object.keys(this.UPDATE_FWRULE_FIELDS).forEach(function (attr) {
@@ -2145,6 +2147,7 @@ function getFirewallRule(id, cb) {
 // <updatable account field> -> <expected typeof>
 CloudApi.prototype.UPDATE_FWRULE_FIELDS = {
     enabled: 'boolean',
+    log: 'boolean',
     rule: 'string',
     description: 'string'
 };
@@ -2160,6 +2163,7 @@ CloudApi.prototype.UPDATE_FWRULE_FIELDS = {
  *      - {UUID} id: The fwrule id. Required.
  *      - {String} rule: The fwrule text. Required.
  *      - {Boolean} enabled: Optional.
+ *      - {Boolean} log: Optional.
  *      - {String} description: Description of the rule. Optional.
  * @param {Function} callback of the form `function (err, fwrule, res)`
  */
@@ -2169,6 +2173,7 @@ function updateFirewallRule(opts, cb) {
     assert.uuid(opts.id, 'opts.id');
     assert.string(opts.rule, 'opts.rule');
     assert.optionalBool(opts.enabled, 'opts.enabled');
+    assert.optionalBool(opts.log, 'opts.log');
     assert.optionalString(opts.description, 'opts.description');
     assert.func(cb, 'cb');
 
diff --git a/lib/do_fwrule/do_create.js b/lib/do_fwrule/do_create.js
index 8405f84..f684cdb 100644
--- a/lib/do_fwrule/do_create.js
+++ b/lib/do_fwrule/do_create.js
@@ -5,14 +5,12 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  *
  * `triton fwrule create ...`
  */
 
 var assert = require('assert-plus');
-var format = require('util').format;
-var vasync = require('vasync');
 
 var common = require('../common');
 var errors = require('../errors');
@@ -21,6 +19,7 @@ var errors = require('../errors');
 function do_create(subcmd, opts, args, cb) {
     assert.optionalString(opts.description, 'opts.description');
     assert.optionalBool(opts.disabled, 'opts.disabled');
+    assert.optionalBool(opts.log, 'opts.log');
     assert.func(cb, 'cb');
 
     if (opts.help) {
@@ -38,9 +37,15 @@ function do_create(subcmd, opts, args, cb) {
     var createOpts = {
         rule: args[0]
     };
+
     if (!opts.disabled) {
         createOpts.enabled = true;
     }
+
+    if (typeof (opts.log) !== undefined) {
+        createOpts.log = opts.log;
+    }
+
     if (opts.description) {
         createOpts.description = opts.description;
     }
@@ -76,6 +81,12 @@ do_create.options = [
         type: 'bool',
         help: 'JSON stream output.'
     },
+    {
+        names: ['log', 'l'],
+        type: 'bool',
+        help: 'Enable logging of new TCP connections or new other-protocol ' +
+            'sessions matching this rule.'
+    },
     {
         names: ['disabled', 'd'],
         type: 'bool',
@@ -95,6 +106,7 @@ do_create.synopses = ['{{name}} {{cmd}} [OPTIONS] RULE-TEXT'];
 
 do_create.help = [
     /* BEGIN JSSTYLED */
+    /* eslint-disable */
     'Create a firewall rule.',
     '',
     '{{usage}}',
@@ -111,6 +123,7 @@ do_create.help = [
     // https://github.com/joyent/sdc-fwrule/blob/master/docs/examples.md
     // or docs.jo Cloud Firewall examples? What link? Ditto in parent.
     /* END JSSTYLED */
+    /* eslint-enable */
 ].join('\n');
 
 do_create.helpOpts = {
diff --git a/lib/do_fwrule/do_list.js b/lib/do_fwrule/do_list.js
index 78d6f23..fa8819b 100644
--- a/lib/do_fwrule/do_list.js
+++ b/lib/do_fwrule/do_list.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  *
  * `triton fwrule list ...`
  */
@@ -17,8 +17,8 @@ var common = require('../common');
 var errors = require('../errors');
 
 
-var COLUMNS_DEFAULT = 'shortid,enabled,global,rule';
-var COLUMNS_LONG = 'id,enabled,global,rule,description';
+var COLUMNS_DEFAULT = 'shortid,enabled,global,log,rule';
+var COLUMNS_LONG = 'id,enabled,global,log,rule,description';
 var SORT_DEFAULT = 'rule';
 
 
diff --git a/lib/do_fwrule/do_update.js b/lib/do_fwrule/do_update.js
index 99f1304..5e1c907 100644
--- a/lib/do_fwrule/do_update.js
+++ b/lib/do_fwrule/do_update.js
@@ -5,20 +5,19 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  *
  * `triton fwrule update ...`
  */
 
-var assert = require('assert-plus');
 var format = require('util').format;
 var fs = require('fs');
 var vasync = require('vasync');
 
 var common = require('../common');
 var errors = require('../errors');
-var UPDATE_FWRULE_FIELDS
-    = require('../cloudapi2').CloudApi.prototype.UPDATE_FWRULE_FIELDS;
+var UPDATE_FWRULE_FIELDS =
+    require('../cloudapi2').CloudApi.prototype.UPDATE_FWRULE_FIELDS;
 
 
 function do_update(subcmd, opts, args, cb) {
diff --git a/lib/do_instance/do_fwrule/do_list.js b/lib/do_instance/do_fwrule/do_list.js
index 5c01873..d44fc35 100644
--- a/lib/do_instance/do_fwrule/do_list.js
+++ b/lib/do_instance/do_fwrule/do_list.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  *
  * `triton instance fwrules ...`
  */
@@ -17,8 +17,8 @@ var common = require('../../common');
 var errors = require('../../errors');
 
 
-var COLUMNS_DEFAULT = 'shortid,enabled,global,rule';
-var COLUMNS_LONG = 'id,enabled,global,rule,description';
+var COLUMNS_DEFAULT = 'shortid,enabled,global,log,rule';
+var COLUMNS_LONG = 'id,enabled,global,log,rule,description';
 var SORT_DEFAULT = 'rule';
 
 
diff --git a/lib/do_instance/do_fwrule/index.js b/lib/do_instance/do_fwrule/index.js
index a2d974e..8213c1a 100644
--- a/lib/do_instance/do_fwrule/index.js
+++ b/lib/do_instance/do_fwrule/index.js
@@ -5,6 +5,8 @@
  */
 
 /*
+ * Copyright 2019 Joyent, Inc.
+ *
  * `triton instance fwrule ...`
  */
 
diff --git a/lib/do_instance/do_fwrules.js b/lib/do_instance/do_fwrules.js
index 62d9516..0b3ed72 100644
--- a/lib/do_instance/do_fwrules.js
+++ b/lib/do_instance/do_fwrules.js
@@ -5,6 +5,8 @@
  */
 
 /*
+ * Copyright 2019 Joyent, Inc.
+ *
  * `triton instance fwrules ...` shortcut for
  * `triton instance fwrule list ...`.
  */
diff --git a/lib/tritonapi.js b/lib/tritonapi.js
index 1e561da..dfa3918 100644
--- a/lib/tritonapi.js
+++ b/lib/tritonapi.js
@@ -2663,7 +2663,8 @@ TritonApi.prototype.getFirewallRule = function getFirewallRule(id, cb) {
     } else {
         this.cloudapi.listFirewallRules({}, function (err, fwrules) {
             if (err) {
-                return cb(err);
+                cb(err);
+                return;
             }
 
             var shortIdMatches = fwrules.filter(function (fwrule) {
@@ -2764,6 +2765,7 @@ function listFirewallRuleInstances(opts, cb) {
  *      - {String} id: The fwrule ID, or short ID. Required.
  *      - {String} rule: The fwrule text. Optional.
  *      - {Boolean} enabled: Default to false. Optional.
+ *      - {Boolean} log: Default to false. Optional.
  *      - {String} description: Description of the rule. Optional.
  *      At least one of the fields must be provided.
  * @param {Function} callback `function (err, fwrule, res)`
@@ -2773,10 +2775,12 @@ TritonApi.prototype.updateFirewallRule = function updateFirewallRule(opts, cb) {
     assert.string(opts.id, 'opts.id');
     assert.optionalString(opts.rule, 'opts.rule');
     assert.optionalBool(opts.enabled, 'opts.enabled');
+    assert.optionalBool(opts.log, 'opts.log');
     assert.optionalString(opts.description, 'opts.description');
     assert.ok(opts.rule !== undefined || opts.enabled !== undefined ||
-        opts.description !== undefined, 'at least one of opts.rule, '
-        + 'opts.enabled, or opts.description is required');
+        opts.description !== undefined || opts.log !== undefined,
+        'at least one of opts.rule, opts.enabled, opts.log or ' +
+        'opts.description is required');
     assert.func(cb, 'cb');
 
     var self = this;
diff --git a/package.json b/package.json
index 8dc4921..d2326c1 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "triton",
   "description": "Joyent Triton CLI and client (https://www.joyent.com/triton)",
-  "version": "7.1.1",
+  "version": "7.2.0",
   "author": "Joyent (joyent.com)",
   "homepage": "https://github.com/joyent/node-triton",
   "dependencies": {
diff --git a/test/integration/cli-fwrules.test.js b/test/integration/cli-fwrules.test.js
index 312dec3..f772139 100644
--- a/test/integration/cli-fwrules.test.js
+++ b/test/integration/cli-fwrules.test.js
@@ -28,6 +28,7 @@ var INST_ALIAS = f('nodetritontest-fwrules-%s', os.hostname());
 var testOpts = {
     skip: !h.CONFIG.allowWriteActions && 'requires config.allowWriteActions'
 };
+var LIST_RE = /ID\s+ENABLED\s+GLOBAL\s+LOG\s+RULE\s+DESCRIPTION/;
 
 // --- Tests
 
@@ -44,8 +45,10 @@ test('triton fwrule', testOpts, function (suite) {
 
     suite.test('  setup: triton create', function (t) {
         h.createTestInst(t, INST_ALIAS, {}, function onInst(err2, instId) {
-            if (h.ifErr(t, err2, 'triton instance create'))
-                return t.end();
+            if (h.ifErr(t, err2, 'triton instance create')) {
+                t.end();
+                return;
+            }
 
             INST = instId;
             RULE = RULE.replace('$id', INST);
@@ -58,10 +61,14 @@ test('triton fwrule', testOpts, function (suite) {
     suite.test('  triton fwrule create --disabled', function (t) {
         var cmd = f('fwrule create -d "%s"', RULE);
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule create --disabled'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule create --disabled')) {
+                t.end();
+                return;
+            }
+            /* eslint-disable max-len */
             /* JSSTYLED */
             var expected = /^Created firewall rule ([a-f0-9-]{36}) \(disabled\)$/m;
+            /* eslint-enable */
             var match = expected.exec(stdout);
             t.ok(match, f('stdout matches %s: %j', expected, stdout));
 
@@ -77,22 +84,27 @@ test('triton fwrule', testOpts, function (suite) {
         var cmd = 'fwrule get ' + ID;
 
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule get'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule get')) {
+                t.end();
+                return;
+            }
 
             var obj = JSON.parse(stdout);
             t.equal(obj.rule, RULE, 'fwrule rule is correct');
             t.equal(obj.enabled, false, 'fwrule is disabled');
+            t.equal(obj.log, false, 'fwrule is not logging');
             t.end();
         });
     });
 
     suite.test('  triton fwrule create', function (t) {
-        var cmd = f('fwrule create -D "%s" "%s"', DESC, RULE);
+        var cmd = f('fwrule create -D "%s" "%s" --log', DESC, RULE);
 
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule create'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule create')) {
+                t.end();
+                return;
+            }
 
             /* JSSTYLED */
             var expected = /^Created firewall rule ([a-f0-9-]{36})$/m;
@@ -111,13 +123,16 @@ test('triton fwrule', testOpts, function (suite) {
         var cmd = 'fwrule get ' + ID;
 
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule get'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule get')) {
+                t.end();
+                return;
+            }
 
             var obj = JSON.parse(stdout);
             t.equal(obj.rule, RULE, 'fwrule rule is correct');
             t.equal(obj.description, DESC, 'fwrule was properly created');
             t.equal(obj.enabled, true, 'fwrule enabled defaults to true');
+            t.equal(obj.log, true, 'fwrule log is to true');
             t.end();
         });
     });
@@ -126,8 +141,10 @@ test('triton fwrule', testOpts, function (suite) {
         var cmd = 'fwrule enable ' + ID;
 
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule enable'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule enable')) {
+                t.end();
+                return;
+            }
 
             t.ok(stdout.match('Enabled firewall rule ' + ID));
 
@@ -139,8 +156,10 @@ test('triton fwrule', testOpts, function (suite) {
         var cmd = 'fwrule disable ' + ID;
 
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule disable'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule disable')) {
+                t.end();
+                return;
+            }
 
             t.ok(stdout.match('Disabled firewall rule ' + ID));
 
@@ -152,8 +171,10 @@ test('triton fwrule', testOpts, function (suite) {
         var cmd = 'fwrule update ' + ID + ' rule="' + RULE2 + '"';
 
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule update'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule update')) {
+                t.end();
+                return;
+            }
 
             t.ok(stdout.match('Updated firewall rule ' + ID +
                  ' \\(fields: rule\\)'));
@@ -162,13 +183,31 @@ test('triton fwrule', testOpts, function (suite) {
         });
     });
 
+    suite.test('  triton fwrule update log', function (t) {
+        var cmd = 'fwrule update ' + ID + ' log=false';
+
+        h.triton(cmd, function (err, stdout, stderr) {
+            if (h.ifErr(t, err, 'triton fwrule update log')) {
+                t.end();
+                return;
+            }
+
+            t.ok(stdout.match('Updated firewall rule ' + ID +
+                 ' \\(fields: log\\)'));
+
+            t.end();
+        });
+    });
+
     suite.test('  triton fwrule list', function (t) {
         h.triton('fwrule list -l', function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule list'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule list')) {
+                t.end();
+                return;
+            }
 
             var rules = stdout.split('\n');
-            t.ok(rules[0].match(/ID\s+ENABLED\s+GLOBAL\s+RULE\s+DESCRIPTION/));
+            t.ok(rules[0].match(LIST_RE));
             rules.shift();
 
             t.ok(rules.length >= 1, 'triton fwrule list expected fwrule num');
@@ -185,11 +224,13 @@ test('triton fwrule', testOpts, function (suite) {
 
     suite.test('  triton fwrules', function (t) {
         h.triton('fwrules -l', function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule list'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule list')) {
+                t.end();
+                return;
+            }
 
             var rules = stdout.split('\n');
-            t.ok(rules[0].match(/ID\s+ENABLED\s+GLOBAL\s+RULE\s+DESCRIPTION/));
+            t.ok(rules[0].match(LIST_RE));
             rules.shift();
 
             t.ok(rules.length >= 1, 'triton fwrule list expected fwrule num');
@@ -206,8 +247,10 @@ test('triton fwrule', testOpts, function (suite) {
 
     suite.test('  triton fwrule instances', function (t) {
         h.triton('fwrule instances -l ' + ID, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule instances'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule instances')) {
+                t.end();
+                return;
+            }
 
             var machines = stdout.split('\n').filter(function (machine) {
                 return machine !== '';
@@ -215,8 +258,10 @@ test('triton fwrule', testOpts, function (suite) {
             t.ok(machines[0].match(/ID\s+NAME\s+IMG\s+BRAND/));
             machines.shift();
 
-            if (!INST)
-                return t.end();
+            if (!INST) {
+                t.end();
+                return;
+            }
 
             t.equal(machines.length, 1, 'triton fwrule instances expected ' +
                     'num machines');
@@ -234,11 +279,13 @@ test('triton fwrule', testOpts, function (suite) {
 
     suite.test('  triton instance fwrules', function (t) {
         h.triton('instance fwrules -l ' + INST, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule list'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule list')) {
+                t.end();
+                return;
+            }
 
             var rules = stdout.split('\n');
-            t.ok(rules[0].match(/ID\s+ENABLED\s+GLOBAL\s+RULE\s+DESCRIPTION/));
+            t.ok(rules[0].match(LIST_RE));
             rules.shift();
 
             t.ok(rules.length >= 1, 'triton fwrule list expected fwrule num');
@@ -256,11 +303,13 @@ test('triton fwrule', testOpts, function (suite) {
     suite.test('  triton instance fwrule list', function (t) {
         h.triton('instance fwrule list -l ' + INST,
             function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule list'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule list')) {
+                t.end();
+                return;
+            }
 
             var rules = stdout.split('\n');
-            t.ok(rules[0].match(/ID\s+ENABLED\s+GLOBAL\s+RULE\s+DESCRIPTION/));
+            t.ok(rules[0].match(LIST_RE));
             rules.shift();
 
             t.ok(rules.length >= 1, 'triton fwrule list expected fwrule num');
@@ -278,8 +327,10 @@ test('triton fwrule', testOpts, function (suite) {
     suite.test('  triton fwrule delete', function (t) {
         var cmd = 'fwrule delete ' + ID + ' --force';
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton fwrule delete'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton fwrule delete')) {
+                t.end();
+                return;
+            }
 
             t.ok(stdout.match('Deleted rule ' + ID + ''), 'rule deleted');
 
@@ -290,15 +341,19 @@ test('triton fwrule', testOpts, function (suite) {
     suite.test('  triton instance enable-firewall', function (t) {
         var cmd = 'instance enable-firewall ' + INST + ' -w';
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton instance enable-firewall'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton instance enable-firewall')) {
+                t.end();
+                return;
+            }
 
             t.ok(stdout.match('Enabled firewall for instance "' + INST + '"'),
                  'firewall enabled');
 
             h.triton('instance get -j ' + INST, function (err2, stdout2) {
-                if (h.ifErr(t, err2, 'triton instance get'))
-                    return t.end();
+                if (h.ifErr(t, err2, 'triton instance get')) {
+                    t.end();
+                    return;
+                }
 
                 var inst = JSON.parse(stdout2);
                 t.equal(inst.firewall_enabled, true);
@@ -311,15 +366,19 @@ test('triton fwrule', testOpts, function (suite) {
     suite.test('  triton instance disable-firewall', function (t) {
         var cmd = 'instance disable-firewall ' + INST + ' -w';
         h.triton(cmd, function (err, stdout, stderr) {
-            if (h.ifErr(t, err, 'triton instance disable-firewall'))
-                return t.end();
+            if (h.ifErr(t, err, 'triton instance disable-firewall')) {
+                t.end();
+                return;
+            }
 
             t.ok(stdout.match('Disabled firewall for instance "' + INST + '"'),
                  'firewall disabled');
 
             h.triton('instance get -j ' + INST, function (err2, stdout2) {
-                if (h.ifErr(t, err2, 'triton instance get'))
-                    return t.end();
+                if (h.ifErr(t, err2, 'triton instance get')) {
+                    t.end();
+                    return;
+                }
 
                 var inst = JSON.parse(stdout2);
                 t.equal(inst.firewall_enabled, false);
-- 
2.21.0

